### PR TITLE
Add api layer for api based authentication

### PR DESCRIPTION
**Pull request by janakamarasena**
API layer for https://github.com/wso2/product-is/issues/15684

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        String authenticatorId = authRequest.getSelectedAuthenticator().getAuthenticatorId();
+        if (authenticatorId != null) {
+            String decodedAuthenticatorId = base64URLDecode(authenticatorId);
+            String[] authenticatorIdSplit = decodedAuthenticatorId.split(AUTHENTICATOR_IDP_SPLITTER);
+
+            if (authenticatorIdSplit.length == 2) {
+                params.put(AUTHENTICATOR, new String[]{authenticatorIdSplit[0]});
+                params.put(IDP, new String[]{authenticatorIdSplit[1]});
+            } else {
+                throw new AuthServiceClientException("Authenticator id is not in the correct format");
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnEndpoint.java:148]: Let's print the authenticator id.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnEndpoint.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+                default:
+                    throw new AuthServiceException("Unknown flow status: " + authServiceResponse.getFlowStatus());
+            }
+
+        } catch (AuthServiceClientException e) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Client error while handling authentication request", e);
+            }
+            return buildOAuthInvalidRequestError(e.getMessage());
+        } catch (AuthServiceException | InvalidRequestParentException | URISyntaxException e) {
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnEndpoint.java:103]: Is  InvalidRequestParentException  a server error? The name of the class sounds different.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+/**
+ * Class containing the REST API for API based authentication.
+ */
+@Path("/authn")
+public class ApiAuthnEndpoint {
+
+    private final AuthenticationService authenticationService = new AuthenticationService();
+    private final OAuth2AuthzEndpoint oAuth2AuthzEndpoint = new OAuth2AuthzEndpoint();
+    private static final String AUTHENTICATOR_IDP_SPLITTER = ":";
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnEndpoint.java:67]: Better move this constants to oAuthConstants

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnHandler.java [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+        return param;
+    }
+
+    private List<Link> buildLinks() throws AuthServiceException {
+
+        List<Link> links = new ArrayList<>();
+        Link authnEpLink = new Link();
+        authnEpLink.setName("authentication");
+        String endpoint = "/oauth2/authn";
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnHandler.java:167]: Let's define these as constants in OAuthConstants.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 5]
```diff
@@ -4263,4 +4296,84 @@ private boolean isFapiConformant(String clientId) throws InvalidRequestException
...
+        }
+
+        return StringUtils.equals(OAuthConstants.ResponseModes.DIRECT,
+                oAuthMessage.getRequest().getParameter(RESPONSE_MODE));
+    }
+
+    private boolean isApiBasedAuthenticationFlow(OAuth2Parameters oAuth2Parameters) {
+
+        if (oAuth2Parameters == null) {
+            log.debug("OAuth2Parameters is null. Returning false for isApiBasedAuthenticationFlow check.");
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:4314]: if (log.isDebugEnabled()) {

Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: fixed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnHandler.java (hunk 4 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnHandler.java]: fixed. Moved to class level constants as currently these are not used in any other place

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnEndpoint.java (hunk 3 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnEndpoint.java]: fixed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnEndpoint.java (hunk 2 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnEndpoint.java]: its a client error. fixed.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnEndpoint.java (hunk 1 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/api/auth/ApiAuthnEndpoint.java]: fixed

**Review by DMHP — DISMISSED** [review]

**Review by DMHP — APPROVED** [review]

**Pull request merged**